File: cpp/tests/test_bachelier.cpp
// Tests for the Bachelier (normal) model: price, implied vol, parity.

#include "irm/bachelier.hpp"

#include <cmath>
#include <limits>
#include <stdexcept>

#include <gtest/gtest.h>

namespace {

TEST(Bachelier, AtmClosedForm) {
    // ATM: price = A sigma sqrt(T) / sqrt(2 pi).
    const double a = 0.98, vol = 0.005, t = 1.0;
    const double p = irm::bachelier_price(-0.002, -0.002, t, vol, a, true);
    EXPECT_NEAR(p, a * vol * std::sqrt(t) / std::sqrt(2.0 * M_PI), 1e-17);
    EXPECT_NEAR(irm::bachelier_price(-0.002, -0.002, t, vol, a, false), p, 1e-17);
    EXPECT_NEAR(irm::bachelier_implied_vol(p, -0.002, -0.002, t, a, true), vol, 1e-14);
}

TEST(Bachelier, ImpliedVolRoundTripGrid) {
    const double t = 2.5, annuity = 3.7;
    for (double forward : {-0.01, 0.0, 0.02, 0.05}) {
        for (double strike : {-0.02, 0.0, 0.03}) {
            for (double vol : {0.001, 0.0075, 0.02}) {
                for (bool payer : {true, false}) {
                    const double p = irm::bachelier_price(forward, strike, t, vol, annuity, payer);
                    const double iv = irm::bachelier_implied_vol(p, forward, strike, t, annuity, payer);
                    const double d = std::fabs(forward - strike) / (vol * std::sqrt(t));
                    if (d < 4.0) {
                        EXPECT_NEAR(iv, vol, 1e-12) << forward << " " << strike << " " << vol;
                    } else {
                        // Time value below rounding: only the price round-trips.
                        EXPECT_GE(iv, 0.0);
                        EXPECT_LE(iv, vol * 1.001);
                    }
                    EXPECT_NEAR(irm::bachelier_price(forward, strike, t, iv, annuity, payer), p, 1e-15);
                }
            }
        }
    }
}

TEST(Bachelier, PayerReceiverParityAndMonotonicity) {
    const double f = 0.03, k = 0.025, t = 2.0, a = 4.2;
    for (double vol : {0.0, 0.002, 0.0075, 0.03}) {
        const double payer = irm::bachelier_price(f, k, t, vol, a, true);
        const double receiver = irm::bachelier_price(f, k, t, vol, a, false);
        EXPECT_NEAR(payer - receiver, a * (f - k), 1e-15);
    }
    double prev = -1.0;
    for (double vol : {0.001, 0.002, 0.005, 0.01, 0.02}) {
        const double p = irm::bachelier_price(f, k, t, vol, a, true);
        EXPECT_GT(p, prev);
        prev = p;
    }
}

TEST(Bachelier, IntrinsicAndBelowIntrinsic) {
    const double f = 0.03, k = 0.025, t = 2.0, a = 4.2;
    const double intrinsic = a * (f - k);
    EXPECT_EQ(irm::bachelier_price(f, k, t, 0.0, a, true), intrinsic);
    EXPECT_EQ(irm::bachelier_price(f, k, t, 0.0, a, false), 0.0);
    EXPECT_EQ(irm::bachelier_implied_vol(intrinsic, f, k, t, a, true), 0.0);
    EXPECT_EQ(irm::bachelier_implied_vol(0.0, f, k, t, a, false), 0.0);
    EXPECT_THROW(irm::bachelier_implied_vol(intrinsic - 1e-6, f, k, t, a, true),
                 std::invalid_argument);
    EXPECT_THROW(irm::bachelier_implied_vol(-1e-9, f, k, t, a, false), std::invalid_argument);
}

TEST(Bachelier, Validation) {
    const double nan = std::nan("");
    EXPECT_THROW(irm::bachelier_price(nan, 0.02, 1.0, 0.01, 1.0), std::invalid_argument);
    EXPECT_THROW(irm::bachelier_price(0.02, 0.02, 0.0, 0.01, 1.0), std::invalid_argument);
    EXPECT_THROW(irm::bachelier_price(0.02, 0.02, 1.0, -0.01, 1.0), std::invalid_argument);
    EXPECT_THROW(irm::bachelier_price(0.02, 0.02, 1.0, 0.01, 0.0), std::invalid_argument);
    EXPECT_THROW(irm::bachelier_implied_vol(nan, 0.02, 0.02, 1.0, 1.0), std::invalid_argument);
    EXPECT_THROW(irm::bachelier_implied_vol(0.01, 0.02, 0.02,
                                            std::numeric_limits<double>::infinity(), 1.0),
                 std::invalid_argument);
    EXPECT_THROW(irm::bachelier_implied_vol(1e300, 0.02, 0.02, 1.0, 1.0), std::invalid_argument);
}

}  // namespace

File: python/tests/test_bachelier.py
"""Bachelier (normal) model: price, implied vol, parity, validation."""

from __future__ import annotations

import math

import pytest

from irm import bachelier_implied_vol, bachelier_price


def test_atm_closed_form() -> None:
    # ATM: price = A sigma sqrt(T) / sqrt(2 pi).
    a, vol, t = 0.98, 0.005, 1.0
    p = bachelier_price(-0.002, -0.002, t, vol, a, True)
    assert p == pytest.approx(a * vol * math.sqrt(t) / math.sqrt(2.0 * math.pi), abs=1e-17)
    assert bachelier_price(-0.002, -0.002, t, vol, a, False) == pytest.approx(p, abs=1e-17)
    assert bachelier_implied_vol(p, -0.002, -0.002, t, a, True) == pytest.approx(vol, abs=1e-14)


@pytest.mark.parametrize("forward", [-0.01, 0.0, 0.02, 0.05])
@pytest.mark.parametrize("strike", [-0.02, 0.0, 0.03])
@pytest.mark.parametrize("vol", [0.001, 0.0075, 0.02])
@pytest.mark.parametrize("payer", [True, False])
def test_implied_vol_round_trip_grid(forward: float, strike: float, vol: float, payer: bool) -> None:
    t, annuity = 2.5, 3.7
    p = bachelier_price(forward, strike, t, vol, annuity, payer)
    iv = bachelier_implied_vol(p, forward, strike, t, annuity, payer)
    d = abs(forward - strike) / (vol * math.sqrt(t))
    if d < 4.0:
        # Time value resolvable in double precision: the vol round-trips.
        assert iv == pytest.approx(vol, abs=1e-12)
    else:
        # Deep in/out of the money the time value is below rounding: the
        # implied vol is whatever reprices the quote, and the *price* must
        # round-trip exactly.
        assert 0.0 <= iv <= vol * 1.001
    assert bachelier_price(forward, strike, t, iv, annuity, payer) == pytest.approx(p, abs=1e-15)


def test_payer_receiver_parity_and_monotonicity() -> None:
    f, k, t, a = 0.03, 0.025, 2.0, 4.2
    for vol in [0.0, 0.002, 0.0075, 0.03]:
        payer = bachelier_price(f, k, t, vol, a, True)
        receiver = bachelier_price(f, k, t, vol, a, False)
        assert payer - receiver == pytest.approx(a * (f - k), abs=1e-15)
    vols = [0.001, 0.002, 0.005, 0.01, 0.02]
    prices = [bachelier_price(f, k, t, v, a, True) for v in vols]
    assert all(b > x for x, b in zip(prices[:-1], prices[1:]))


def test_intrinsic_and_below_intrinsic() -> None:
    f, k, t, a = 0.03, 0.025, 2.0, 4.2
    intrinsic = a * (f - k)
    assert bachelier_price(f, k, t, 0.0, a, True) == intrinsic
    assert bachelier_price(f, k, t, 0.0, a, False) == 0.0
    assert bachelier_implied_vol(intrinsic, f, k, t, a, True) == 0.0
    assert bachelier_implied_vol(0.0, f, k, t, a, False) == 0.0
    with pytest.raises(ValueError, match="intrinsic"):
        bachelier_implied_vol(intrinsic - 1e-6, f, k, t, a, True)
    with pytest.raises(ValueError, match="intrinsic"):
        bachelier_implied_vol(-1e-9, f, k, t, a, False)


def test_validation() -> None:
    nan = float("nan")
    with pytest.raises(ValueError):
        bachelier_price(nan, 0.02, 1.0, 0.01, 1.0)
    with pytest.raises(ValueError):
        bachelier_price(0.02, 0.02, 0.0, 0.01, 1.0)
    with pytest.raises(ValueError):
        bachelier_price(0.02, 0.02, 1.0, -0.01, 1.0)
    with pytest.raises(ValueError):
        bachelier_price(0.02, 0.02, 1.0, 0.01, 0.0)
    with pytest.raises(ValueError):
        bachelier_implied_vol(nan, 0.02, 0.02, 1.0, 1.0)
    with pytest.raises(ValueError):
        bachelier_implied_vol(0.01, 0.02, 0.02, float("inf"), 1.0)
    with pytest.raises(ValueError, match="too large"):
        bachelier_implied_vol(1e300, 0.02, 0.02, 1.0, 1.0)
